### DISCUSSION TITLE
Welcome to the Royal Kludge R65 Discussions!

### DISCUSSION THREAD

**Discussion by irfanjmdn** [irfanjmdn/r65#1] [General]
To all developers who want to contribute, any and all help would be greatly appreciated! Feel free to submit improvements, fixes, or suggestions.

**Comment by roperscrossroads** [comment]
Thanks for all the work you've put in to this project!

I bought two of the black ansi wired keyboards from amazon recently ([link](https://www.amazon.com/dp/B0CNP1BG1W)) and mine have a different productId than yours. I'm a bit confused about what I've received. I hope they have the same processor and I can just program them using your code. I'm going to investigate more in the next day or two.  I forked your code ([link](https://github.com/roperscrossroads/r65-variants)) and added some [screenshots](https://github.com/irfanjmdn/r65/commit/09eb9bef08750b9f46d6f15c0f4328f405642a2f) of diffs of your **RK R65 Layout for Via.json** vs the current VIA json files on Royal Kludge website. It seems like a bit of a mess. They have reused names and IDs for different products/variants.

The VIA json file appears to be the same except for the name and productId.

```
[nix-shell:~/qmk/r65-variants]$ lsusb
Bus 001 Device 009: ID 342d:e481 Hangsheng RK-R65
```

I also added some of this info over on @iamdanielv's repo [link](https://github.com/iamdanielv/kb_rk_r65/issues/1) so if other people have issues they can have a look at my repo. I'll add more info in my fork once I figure out the differences.

**Reply by BabyJenn** [reply]
same as me. i have R65 keyboard with PID:e481 and i tried @irfanjmdn firmware and the the PID changed to e453 but everything going well. and the keyboard didnt break. and i tried to find the original firmware with PID:e481 but i couldnt find it. maybe u have the original firmware with PID:e481?

**Reply by roperscrossroads** [reply]
@BabyJenn I replied on another thread which I think you probably saw. Sorry, I don't have the original firmware with PID e481

For other people that are trying to track down info on the R65, here are some links:

- https://github.com/irfanjmdn/r65 (this repo)
- https://github.com/iamdanielv/kb_rk_r65
- https://github.com/iamdanielv/kb_rk_r65/issues/1
- https://github.com/roperscrossroads/r65-variants (my research)
- https://www.youtube.com/watch?v=OxRa_ZbAYyQ (disassembly video)

**Comment by BabyJenn** [comment]
i tried all ur firmware on my R65 keyboard and all firmware worked well. and i decided to chose the defaut firmware with SOCD optional. but the problem is when i use the default firmware with SOCD optional, the shortcut of mute,next, previous, calculator etc didnt work. i check on VIA to see the shortcut, and shortcut appear on keyboard layout but didnt work. can you help me with that?

**Reply by irfanjmdn** [reply]
Hello, I wouldn't recommend flashing the default firmware with SOCD as it doesn't have VIA support. You could flash the VIA + SOCD version as it has more features, and the shortcuts can be controlled thru VIA.

**Reply by BabyJenn** [reply]
yeah i used that firmware VIA with SOCD but the shortcut wont work.

**Reply by irfanjmdn** [reply]
> yeah i used that firmware VIA with SOCD but the shortcut wont work.

Ohh that's simple, just go to https://usevia.app/ and switch MO(1) to MO(5), since all of the music shortcuts are on that layer. You can refer to the video I've attached.

https://github.com/user-attachments/assets/a2aebb41-1152-4cbc-b605-75a27af99093

**Reply by BabyJenn** [reply]
ohh i've got itt.. thanks man. i like ur all firmware. I truly appreciate your help

**Reply by BabyJenn** [reply]
I tried from the video you attached. it's work well, but it's not like the original but it can be a alternative. because the original you should go with FN+L CTRL first and then after that you in on the different layer and then press FN with Number so you can go for music shortcut. i'll show you the manual book 
[RK R65 Manual Book.pdf](https://github.com/user-attachments/files/18688514/RK.R65.Manual.Book.pdf).
and if I switch MO(5) on FN button I cannot turn on/off SOCD toggle..
I hope you understand what I mean.

**Reply by derekjc** [reply]
@irfanjmdn Left control key in layer 1/3 is "CUSTOM(88)". They seem to switch to layer 4/6. Are you aware of how this works?

<img width="506" alt="image" src="https://github.com/user-attachments/assets/5eaab3da-3f53-4a18-a0ca-a08f641d5695" />

**Comment by Edduaardo** [comment]
Thank you for this repo! I got really sad when I received my R65 just to discover that the manufacturer didn't published their firmware. Just flashed it on my R65 today and so far is working wonders. On my next keyboard purchase I will verify beforehand if the firmware is available.
